In [1]:
import pandas as pd
import datacompy as dc

import load_dat
import load_xml

## Data comparaison

The bulk of the code can be found in load_dat.py and load_xml.py along with some comments explaining the code. My intention has been to find a balance between readability, robustness and efficiency.

I have opted to use DataCompy when comparing the two sets of data. I find it easy to use, fast and most importantly it produces an result that gives a good overview.




In [2]:
xml_df = load_xml.load_xml("animals.xml")

In [3]:
dat_df = load_dat.load_dat("animals.dat")

In [4]:
def order_columns(df):
    df = df[['id', 'morph', 'miff', 'xcam', 'classA', 'classB', 'classC']]
    return df

In [5]:
dat_df=order_columns(dat_df)
dat_df

id morph  miff  xcam            classA          classB  \
0      10010   236   442  1462               NaN        314, 819   
1      10011    16  1852  6394           1426.80      1819, 1507   
2      10012   206  1985  3418             71.36       1779, 125   
3      10013    40   311  7300  1719.17, 1597.71             NaN   
4      10014   426  1435  6549           1088.77             NaN   
...      ...   ...   ...   ...               ...             ...   
88982  98995   418  1706  4898   1742.15, 258.58            1605   
88983  98996   124  1559  9719               NaN      1052, 1961   
88984  98997   146  1484  5864            171.85             NaN   
88985  98998   105  1657  1949               NaN        701, 748   
88986  98999   279  1775  1588            553.75  146, 1314, 521   

                       classC  
0               1157.4, 529.9  
1                      1213.2  
2                      1409.1  
3                      1556.5  
4                154.5, 261.1  
...                       ...  
88982                  1195.8  
88983                     NaN  
88984                   687.2  
88985  1482.6, 1993.5, 1490.9  
88986                     NaN  

[88987 rows x 7 columns]

In [6]:
xml_df=order_columns(xml_df)
xml_df

id morph  miff  xcam            classA          classB  \
0      10010   236   442  1462               NaN        314, 819   
1      10011    16  1852  6394           1426.80      1819, 1507   
2      10012   206  1985  3418             71.36       1779, 125   
3      10013    40   311  7300  1719.17, 1597.71             NaN   
4      10014   426  1435  6549           1088.77             NaN   
...      ...   ...   ...   ...               ...             ...   
88982  98995   418  1706  4898   1742.15, 258.58            1605   
88983  98996   124  1559  9719               NaN      1052, 1961   
88984  98997   146  1484  5864            171.85             NaN   
88985  98998   105  1657  1949               NaN        701, 748   
88986  98999   279  1775  1588            553.75  146, 1314, 521   

                       classC  
0               1157.4, 529.9  
1                      1213.2  
2                      1409.1  
3                      1556.5  
4                154.5, 261.1  
...                       ...  
88982                  1195.8  
88983                     NaN  
88984                   687.2  
88985  1482.6, 1993.5, 1490.9  
88986                     NaN  

[88987 rows x 7 columns]

In [7]:
compared = dc.Compare(
xml_df,
dat_df,
join_columns='id', 
abs_tol=0,
rel_tol=0,
df1_name='XML',
df2_name='DAT' 
)

In [8]:
print(compared.report())

DataComPy Comparison
--------------------

DataFrame Summary
-----------------

  DataFrame  Columns   Rows
0       XML        7  88987
1       DAT        7  88987

Column Summary
--------------

Number of columns in common: 7
Number of columns in XML but not in DAT: 0
Number of columns in DAT but not in XML: 0

Row Summary
-----------

Matched on: id
Any duplicates on match values: No
Absolute Tolerance: 0
Relative Tolerance: 0
Number of rows in common: 88,984
Number of rows in XML but not in DAT: 3
Number of rows in DAT but not in XML: 3

Number of rows with some compared columns unequal: 14
Number of rows with all compared columns equal: 88,970

Column Comparison
-----------------

Number of columns compared with some values unequal: 6
Number of columns compared with all values equal: 1
Total number of values which compare unequal: 17

Columns with Unequal Values or Types
------------------------------------

   Column XML dtype DAT dtype  # Unequal  Max Diff  # Null Diff
3  classa 

## Conclusion 

There is 17 values that differs between the two data sets, namely: 

    id        morph(XML)    morph(DAT)
0  23768         458           454
1  75948         488           489

    id       miff(XML)     miff(DAT)
0  26553        564          562
1  16029        227          237
2  23768       1798         1799

    id       xcam(XML)     xcam(DAT)
0  23768       4035         5035
1  35610       3908         3909

    id       classA(XML)   classA(DAT)
0  93647       324.93          NaN
1  58974      1833.37      1833.97
2  89138       173.14       273.14

    id       classB(XML)   classB(DAT)
0  86036          NaN         1952
1  36809          432          442
2  19817          200          145

    id         classC(XML)               classC(DAT)
0  93647         986.3            986.3, 324.92
1  11738  907.3, 614.4, 271.2     907.3, 604.4, 271.2
2  83183         1931.5                  1931.6
3  80306         1730.1                  1737.1


Some of these might be considered acceptable if the tolerance was higher than zero. In 3 instances a class value was completely missing(86036 and twice in 93647), it does seam likely that one of the values in animal 93647 has been placed in the wrong class.